engine/swarm/secrets.md

```diff
@@ -0,0 +1,812 @@
+    password. You can use another command to generate the password if you
+    choose.
+
+    > **Note**: After you create a secret, you cannot update it. You can only
+    > remove and re-create it, and you cannot remove a secret that a service is
+    > using. However, you can grant or revoke a running service's access to
+    > secrets using `docker service update`. If you need the ability to update a
+    > secret, consider adding a version component to the secret name, so that you
+    > can later add a new version, update the service to use it, then remove the
+    > old version.
+
+    ```bash
+    $ openssl rand -base64 20 | docker secret create mysql_password
+
+    tmev4fa4ovwkmamcz97sfzlsz
+    ```
+
+    The value returned is not the password, but the ID (digest) of the secret.
+    List the secrets managed by Docker using `docker secret ls`:
+
+    ```bash
+    $ docker secret ls
+
+    ID                          NAME                CREATED             UPDATED
+    tmev4fa4ovwkmamcz97sfzlsz   mysql_password      41 seconds ago      41 seconds ago
+    ```
+
+    The secret is stored in the encrypted WAL logs for the swarm.
+
+2.  Create a user-defined overlay network which will be used for communication
+    between the MySQL and WordPress services. There is no need to expose the
+    MySQL service to any external host or container.
+
+    ```bash
+    $ docker network create -d overlay mysql_private
+    ```
+
+2.  Create the MySQL service. The MySQL service will have the following
+    characteristics:
+
+    - Because the scale is set to `1`, only a single MySQL task runs.
+      Load-balancing MySQL is left as an exercise to the reader and involves
+      than just scaling the service.
+    - Only reachable by other containers on the `mysql_private` network.
+    - Uses the volume `mydata` to store the MySQL data, so that it Persists
+      across restarts to the `mysql` service.
+    - The secret is mounted in a `tmpfs` filesystem at `/run/secrets/mysql_password`
+      and is not group-or-world-readable. It is never exposed as an environment
+      variable, nor can it be committed to an image if the `docker commit`
+      command is run.
+    - Sets the environment variable `MYSQL_ROOT_PASSWORD_FILE` to point to the
+      `mysql` password in the bind-mounted `tempdir` volume. The `mysql` image
+      reads the password string from that file when initializing the database
+      for the first time. Afterward, it is stored in the MySQL database itself.
+
+      ```bash
+      $ docker service create \
+                       --name mysql \
+                       --replicas 1 \
+                       --network mysql_private \
+                       --mount type=volume,source=mydata,destination=/var/lib/mysql \
+                       --secret source=mysql_password,target=mysql_password \
+                       -e MYSQL_ROOT_PASSWORD_FILE="/run/secrets/mysql_password" \
+                      mysql:latest
```

Comment:
You need _two_ secrets here; one for the `root` password and one for the `wordpress` user;

```
docker service create \
  --name mysql \
  --replicas 1 \
  --network=mysql_private \
  --mount type=volume,source=mydata,destination=/var/lib/mysql \
  --secret=mysql_root_password \
  --secret=mysql_password \
  -e MYSQL_ROOT_PASSWORD_FILE="/run/secrets/mysql_root_password" \
  -e MYSQL_PASSWORD_FILE="/run/secrets/mysql_password" \
  -e MYSQL_DATABASE=wordpress \
  -e MYSQL_USER=wordpress \
  mysql
```

here;

- `MYSQL_ROOT_PASSWORD_FILE` sets the path to the secret for the `root` password
- `MYSQL_PASSWORD_FILE ` sets the path to the secret for the regular user's (`wordpress`) password
- `MYSQL_DATABASE` sets the name of the database to create the first time the service is started
- `MYSQL_USER` sets the name of the _user_ to create the first time the service is started